File: samples/Test/Classes/Control/AutoCollectionViewHorizontalTest.cpp

#include "AutoCollectionViewHorizontalTest.h"
#include "FirstViewController.h"


AutoCollectionViewHorizontalTest::AutoCollectionViewHorizontalTest()
{
    this->setTitle("CAAutoCollectionView");
    
    for (int i = 0; i < 60; i++)
    {
        char r = CCRANDOM_0_1() * 255;
        char g = CCRANDOM_0_1() * 255;
        char b = CCRANDOM_0_1() * 255;
        
        m_vColors.push_back(CAColor4B(r, g, b, 255));
    }
}

AutoCollectionViewHorizontalTest::~AutoCollectionViewHorizontalTest()
{

}

void AutoCollectionViewHorizontalTest::viewDidLoad()
{
    m_pAutoCollection = CAAutoCollectionView::createWithLayout(DLayoutFill);
    m_pAutoCollection->setAllowsSelection(true);
    m_pAutoCollection->setAllowsMultipleSelection(true);
    m_pAutoCollection->setOrientation(CAAutoCollectionView::Orientation::Horizontal);
    m_pAutoCollection->setHoriCellInterval(20);
    m_pAutoCollection->setVertCellInterval(20);
    m_pAutoCollection->setReachBoundaryHandOverToSuperview(false);
    this->getView()->addSubview(m_pAutoCollection);
    
    m_pAutoCollection->onCellAtIndexPath(CALLBACK_BIND_3(AutoCollectionViewHorizontalTest::collectionCellAtIndex, this));
    m_pAutoCollection->onNumberOfItemsAtIndexPath(CALLBACK_BIND_1(AutoCollectionViewHorizontalTest::numberOfItemsInSection, this));
    m_pAutoCollection->onCellSizeAtIndexPath(CALLBACK_BIND_2(AutoCollectionViewHorizontalTest::collectionViewCellSizeAtIndexPath, this));
    
    m_pAutoCollection->onDidSelectCellAtIndexPath(CALLBACK_BIND_2(AutoCollectionViewHorizontalTest::collectionViewDidSelectCellAtIndexPath, this));
    
}

void AutoCollectionViewHorizontalTest::viewDidUnload()
{
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

CACollectionViewCell* AutoCollectionViewHorizontalTest::collectionCellAtIndex(const DSize& cellSize, unsigned int section, unsigned int item)
{
    CACollectionViewCell* cell = m_pAutoCollection->dequeueReusableCellWithIdentifier("CrossApp");

    if (cell == NULL)
    {
        cell = CACollectionViewCell::create("CrossApp");
        
        CAView* itemImage = CAView::createWithLayout(DLayoutFill);
        itemImage->setTag(99);
        cell->getContentView()->addSubview(itemImage);
        
        CALabel* itemText = CALabel::createWithLayout(DLayoutFill);
        itemText->setTag(100);
        itemText->setFontSize(29);
        itemText->setTextAlignment(CATextAlignment::Center);
        itemText->setVerticalTextAlignmet(CAVerticalTextAlignment::Center);
        cell->getContentView()->addSubview(itemText);
    }

    CAView* itemImageView = cell->getContentView()->getSubviewByTag(99);
    itemImageView->setColor(m_vColors.at(item));
    CCLog("row = %d", item);
    
    char pos[20] = "";
    sprintf(pos, "(%d,%d)", section, item);
    CALabel* itemText = (CALabel*)cell->getContentView()->getSubviewByTag(100);
    itemText->setText(pos);
    return  cell;
    
}

DSize AutoCollectionViewHorizontalTest::collectionViewCellSizeAtIndexPath(unsigned int section, unsigned int item)
{
    return DSize(230, 230);
}

unsigned int AutoCollectionViewHorizontalTest::numberOfItemsInSection(unsigned int section)
{
    return (unsigned int)m_vColors.size();
}

void AutoCollectionViewHorizontalTest::collectionViewDidSelectCellAtIndexPath(unsigned int section, unsigned int item)
{
    CACollectionViewCell* cell = m_pAutoCollection->cellForRowAtIndexPath(section, item);
    cell->getContentView()->setRotation(-360);
    cell->getContentView()->setScale(0.5f);
    CAViewAnimation::beginAnimations("");
    cell->getContentView()->setRotation(0);
    cell->getContentView()->setScale(1.0f);
    CAViewAnimation::commitAnimations();
}

File: samples/Test_JS/Resources/js/Control/AutoCollectionViewHorizontalTest.js
/**
 * Created by crossApp on 16/8/8.
 */

var AutoCollectionViewHorizontalTest = ca.CAViewController.extend({

    ctor: function () {
        this._super();
    },

    viewDidLoad: function() {

        this.HorizontalcolorArr = new Array();
        for (var i = 0; i < 40; i++)
        {
            var r = Math.floor(Math.random()*255);
            var g = Math.floor(Math.random()*255);
            var b = Math.floor(Math.random()*255);

            this.HorizontalcolorArr.push(ca.CAColor4B.set(r, g, b, 255));
        }
                                                                  

        this.autoCollection = ca.CAAutoCollectionView.createWithLayout(ca.DLayout.fill());
        this.autoCollection.setAllowsSelection(true);
        this.autoCollection.setAllowsMultipleSelection(true);
        this.autoCollection.setOrientation(ca.CAAutoCollectionView.Orientation.Horizontal);
        this.autoCollection.setHoriCellInterval(20);
        this.autoCollection.setVertCellInterval(20);
        this.autoCollection.setReachBoundaryHandOverToSuperview(false);
        this.getView().addSubview(this.autoCollection);
                                                                  
                                                                  
        this.autoCollection.onDidSelectCellAtIndexPath(this.collectionViewDidSelectCellAtIndexPath.bind(this));
        this.autoCollection.onCellAtIndexPath(this.collectionCellAtIndex.bind(this));
        this.autoCollection.onCellSizeAtIndexPath(this.collectionViewSizeForItemAtIndexPath.bind(this));
        this.autoCollection.onNumberOfItemsAtIndexPath(this.numberOfItemsInSection.bind(this));
        this.autoCollection.onNumberOfSections(this.numberOfSections.bind(this));
    },

    collectionViewDidSelectCellAtIndexPath: function(section, item)
    {
        var cell = this.autoCollection.cellForRowAtIndexPath(section, item);
        cell.getContentView().setRotation(-360);
        cell.getContentView().setScale(0.5);
        ca.CAViewAnimation.beginAnimations("");
        cell.getContentView().setRotation(0);
        cell.getContentView().setScale(1.0);
        ca.CAViewAnimation.commitAnimations();
    },
    collectionCellAtIndex: function(cellSize, section, item)
    {
        var cell = this.autoCollection.dequeueReusableCellWithIdentifier("cell");

        if (cell == null)
        {
            cell = ca.CACollectionViewCell.create("cell");

            var itemImage = ca.CAView.createWithLayout(ca.DLayout.fill());
            itemImage.setTag(99);
            cell.getContentView().addSubview(itemImage);

            var itemText = ca.CALabel.createWithLayout(ca.DLayout.fill());
            itemText.setTag(100);
            itemText.setFontSize(29);
            itemText.setTextAlignment(ca.CATextAlignment.Center);
            itemText.setVerticalTextAlignmet(ca.CAVerticalTextAlignment.Center);
            cell.getContentView().addSubview(itemText);
        }

        var itemImageView = cell.getContentView().getSubviewByTag(99);
        itemImageView.setColor(this.HorizontalcolorArr[item]);
        ca.log("row = "+ item);


        var itemText = cell.getContentView().getSubviewByTag(100);
        itemText.setText("("+section+","+item+")");
        return  cell;

    },

    collectionViewSizeForItemAtIndexPath: function( collectionView, section, item)
    {
        return ca.DSize.set(230, 230);
    },

    numberOfItemsInSection: function(collectionView, section)
    {
        return this.HorizontalcolorArr.length;
    },
    numberOfSections: function(collectionView)
    {
        return 1;
    }

});
